File: go/2022/day04/day04.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"
)

var inputFile = flag.String("inputFile", "input.txt", "Relative file path to use as input")

func main() {
	flag.Parse()
	bytes, err := os.ReadFile(*inputFile)
	if err != nil {
		panic("Unable to read the input file")
	}

	input := string(bytes)

	fmt.Printf("AoC 2022, Day4, Part1 solution is: %d \n", part1(input))
	fmt.Printf("AoC 2022, Day4, Part2 solution is: %d \n", part2(input))
}

func part1(input string) int {
	pairsList := parseInput(input)

	total := 0

	for _, pairs := range pairsList {
		if pairs[0].includes(pairs[1]) || pairs[1].includes(pairs[0]) {
			total += 1
		}
	}

	return total
}

func part2(input string) int {
	pairsList := parseInput(input)

	total := 0

	for _, pairs := range pairsList {
		if pairs[0].overlaps(pairs[1]) || pairs[1].overlaps(pairs[0]) {
			total += 1
		}
	}

	return total
}

type pair struct {
	x int
	y int
}

func (p1 *pair) includes(p2 pair) bool {
	return p1.x >= p2.x && p1.x <= p2.y && p1.y >= p2.x && p1.y <= p2.y
}

func (p1 *pair) overlaps(p2 pair) bool {
	return p1.y >= p2.x && p1.y <= p2.y ||
		p1.x >= p2.x && p1.x <= p2.y ||
		p1.x >= p2.x && p1.x <= p2.y && p1.y >= p2.x && p1.y <= p2.y
}

func parseInput(input string) [][2]pair {
	assignments := [][2]pair{}

	for _, line := range strings.Split(input, "\n") {
		pair1 := pair{}
		pair2 := pair{}

		fmt.Sscanf(line, "%d-%d,%d-%d", &pair1.x, &pair1.y, &pair2.x, &pair2.y)

		assignments = append(assignments, [2]pair{pair1, pair2})
	}

	return assignments
}

File: python/2022/day04/day04.py
from typing import Callable, List


def run(input: str, fn: Callable[[List[int], List[int]], bool]) -> int:
    count = 0
    for line in parse_input(input):
        parts = line.split(",")
        assert len(parts) == 2

        elf1 = [int(x) for x in parts[0].split("-")]
        elf2 = [int(x) for x in parts[1].split("-")]

        if fn(elf1, elf2):
            count += 1

    return count


def part1_condition(elf1: List[int], elf2: List[int]) -> bool:
    assert len(elf1) == 2
    assert len(elf2) == 2

    (x1, x2) = elf1
    (y1, y2) = elf2

    return (x1 <= y1 <= y2 <= x2) or (y1 <= x1 <= x2 <= y2)


def part2_condition(elf1: List[int], elf2: List[int]) -> bool:
    assert len(elf1) == 2
    assert len(elf2) == 2

    (x1, x2) = elf1
    (y1, y2) = elf2

    return (
        x1 <= y1 <= y2 <= x2
        or y1 <= x1 < +x2 <= y2
        or x1 <= y1 <= x2
        or x1 <= y2 <= x2
        or y1 <= x1 <= y2
        or y1 <= x2 <= y2
    )


def parse_input(input: str) -> List[str]:
    with open(input) as fn:
        raw_data = fn.read().strip()
        parts = raw_data.split("\n")
    return parts


def main():
    puzzle_input_path = "2022/day04/input.txt"
    print(f"AoC2022 Day4, Part1 solution is: {run(puzzle_input_path, part1_condition)}")
    print(f"AoC2022 Day4, Part2 solution is: {run(puzzle_input_path, part2_condition)}")


if __name__ == "__main__":
    main()
